File: persistence-with-c1-chat/src/app/api/chat/tools.ts
import type { RunnableToolFunctionWithParse } from "openai/lib/RunnableFunction.mjs";
import { z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import type { JSONSchema } from "openai/lib/jsonschema.mjs";

export const tools: RunnableToolFunctionWithParse<{ location: string }>[] = [
  {
    type: "function",
    function: {
      name: "getWeather",
      description: "Get the weather for the given location",
      parse: (input) => {
        console.log("input lkasjdlkfjds", input);
        return JSON.parse(input) as { location: string };
      },
      parameters: zodToJsonSchema(
        z.object({
          location: z.string().describe("location for weather"),
        })
      ) as JSONSchema,
      function: async ({ location }: { location: string }) => {
        const weathers = ["sunny", "cloudy", "rainy", "snowy"];
        const randomWeather =
          weathers[Math.floor(Math.random() * weathers.length)];

        return `The weather in ${location} is ${randomWeather}`;
      },
      strict: true,
    },
  },
];

File: langgraph-with-c1-python/backend/tools.py
import random
from langchain_core.tools import tool
from pydantic import BaseModel, Field

class WeatherInput(BaseModel):
    location: str = Field(description="The location for which to get the weather.")

@tool(args_schema=WeatherInput)
def get_weather(location: str) -> str:
    """Get the current weather for the given location."""
    weathers = ["sunny", "cloudy", "rainy", "snowy", "windy", "foggy"]
    return f"The weather in {location} is currently {random.choice(weathers)}."

# List of tools to be used by the agent
runnable_tools = [get_weather] 
